Break down kcidb.orm module

octonawish-akcodes opened the pull request:
Extracted `Schema` and `Pattern` class and their dependencies and made them as a separate module from kcidb.orm module.
This PR closes #386 
@spbnick Need reviews.

spbnick commented on kcidb/orm/data.py:
```diff
@@ -0,0 +1,302 @@
+from kcidb.orm.__init__ import Type, Relation,_DEFS, _CHECKOUT, _BUILD, _TEST, _INCIDENT, _ISSUE, _ISSUE_CULPRIT, _TEST_ENVIRONMENT
```
This would create a circular dependency, and shouldn't happen.

We need to bring the definitions of these dependencies here as well (as I pointed out in the issue).

Additionally, please do not import from `__init__` directly. This creates a very tight "coupling" and makes the code more fragile.

octonawish-akcodes replied:
Okay, I will do that.

spbnick commented on kcidb/orm/query.py:
```diff
@@ -0,0 +1,1461 @@
+from kcidb.orm.__init__ import re, Type, LIGHT_ASSERTS, textwrap
```
Same here: do not import anything from `__init__` explicitly. Also, definitely do not import common definitions available elsewhere (such as `LIGHT_ASSERTS`, or `textwrap`) via a third-party module.

octonawish-akcodes replied:
Okay sir.

spbnick reviewed (changes requested):
Nice start, Abhishek! Left you a few comments/requests inline. Thanks!

spbnick commented:
Don't hesitate to ask any questions you might have about this, here, or on Slack!

octonawish-akcodes commented on kcidb/orm/query.py:
```diff
@@ -0,0 +1,1461 @@
+from kcidb.orm.__init__ import re, Type, LIGHT_ASSERTS, textwrap
+from kcidb.orm.data import Schema, SCHEMA
+import kcidb.io as io
+import kcidb
```
got it.

octonawish-akcodes commented:
After making the suggested changes, Below is the output of the test cases. All are passing as they were passing before creating the separate modules for `Schema` and ` Pattern`.

![image](https://user-images.githubusercontent.com/76171953/226326475-d9570fec-69b0-4b2f-8681-ee0cfae3c905.png)

octonawish-akcodes commented:
In updated the PR and removed the reformatting, I'll make a separate PR for that if necessary. Now this PR is purely the break down changes to be reviewed @spbnick

spbnick reviewed (changes requested):
Very nice! We're close to the finish line here!

Please deal with the CI errors, and from now on please make sure CI passes before review, as you're now more experienced with the project (please keep ignoring the deployment failures for now).

Please also don't forget to keep rebasing your changes on latest `main` as you work, and before reviews.

Finally, please explain a little what you're doing in the commit message body (take the issue description as a reference).

octonawish-akcodes commented:
Test cases passed, fixes the cyclic imports, and updated the references where necessary. I also added a description of what I am doing and what I have done have a look, sir. This PR is intended to work as we wanted it to, need reviews @spbnick

spbnick commented on kcidb/db/abstract.py:
```diff
@@ -1,7 +1,7 @@
 
 from abc import ABC, abstractmethod
-import kcidb.orm as orm
+from kcidb.orm.query import Pattern
```
Could you instead write:
```python
import kcidb.orm.query
```
And refer to `Pattern` as `kcidb.orm.query.Pattern`? We only have one reference here (so there's not much typing/reading overhead) and it would be good to be clear where it's coming from exactly.

octonawish-akcodes replied:
Oh!!! Okay. Done.

spbnick commented on kcidb/orm/data.py:
```diff
@@ -0,0 +1,478 @@
+"""
+The kcidb.orm.data module defines the data structures and constants used
+by the KCIDB object-relational mapping system to organize Kernel CI report
+data into objects.
```
Let's not put the module name into its description, so we don't forget to update it if we ever rename it. Besides, it's redundant. You know that it's a module, and which module it is when you're reading this. Otherwise this is a pretty good description, thanks! But let's start it with just "The data structures and constants ..." directly.

spbnick commented on kcidb/orm/query.py:
```diff
@@ -0,0 +1,830 @@
+"""
+The kcidb.orm.query module defines the query classes and functions used
+by the KCIDB object-relational mapping system to retrieve Kernel CI report
+data from the database.
```
Same here, let's just write "The query classes ..." and so on. Otherwise, another nice one!

octonawish-akcodes replied:
Done!!!

spbnick reviewed (changes requested):
Ah, very neat and tidy :relaxed: Thank you, Abhishek!

I only have a few minor requests left.

The commit message looks very good too, thanks!

It would be totally perfect if you used the imperative mood in the commit body as well (i.e. wrote "Extract parts ...", instead of "We extracted parts ...", and so on), but the most important part is the mood of the commit subject, and that's fine.

octonawish-akcodes commented:
@spbnick Done all the requested changes, Need reviews.

spbnick commented on kcidb/orm/__init__.py:
```diff
@@ -3,1287 +3,27 @@
 import kcidb.misc
 from kcidb.misc import LIGHT_ASSERTS
+from kcidb.orm.query import Pattern, PatternHelpAction
+from kcidb.orm.data import SCHEMA
```
Could you please import the `query` and `data` modules here, instead of the symbols directly? And refer to the symbols through them, e.g. `query.Pattern` and `data.SCHEMA`. First, this would satisfy #404, and then it would stop these symbols appearing like they're actually still defined here. It's best to always refer to the symbols by their canonical name and definition location, and not use third-party references (e.g. like the `Pattern` imported here would be).

In other words, the symbols you **moved** out of the `kcidb.orm` module should no longer be available there.

This will also require updating more code in other modules, including `main.py` in the top directory, containing the Cloud Function entry points.

octonawish-akcodes replied:
Okay

spbnick commented on kcidb/orm/__init__.py:
```diff
@@ -3,1287 +3,27 @@
 assert set(SCHEMA.types) >= \
     set(k[:-1] for k in io.SCHEMA.graph if k), \
     "OO types are not a superset of I/O types"
 
```
These assertions should go into the `data` submodule, together with the `SCHEMA`, as they're closely tied to it, and are verifying either its pre-conditions, or the `SCHEMA` itself.

octonawish-akcodes replied:
done!!!

spbnick reviewed (changes requested):
Thank you very much Abhishek! This looks very good and satisfies all my previous requests.

However, now that your PR is much neater I was able to notice one small, and one larger problem with this. I left you a couple requests inline.

Thank you.

spbnick commented:
@octonawish-akcodes, I tried reproducing your import problems, and it would go smoother if instead of things like this:
```python
import kcidb.orm.query as query
import kcidb.orm.query as data
```
you would write
```python
from kcidb.orm import query, data
```

spbnick commented:
We don't have much time left for you to clean this up, and we're blocking #379, so I went ahead and did it for you. Please review the patch I added on top of yours, then squash them together, push to PR, and then review the result again.

Main changes I did:
* Remove all unnecessary changes, including whitespace and import changes. It's generally better to minimize churn and do only what's necessary for a particular task (provided it's done well). We can work on unifying and refining imports later.
* Update documentation so it refers to the new location of the moved symbols. Always remember to update the docs.
* Remove the `LOGGER` global from modules that don't need it. Do not add globals unnecessarily.

spbnick reviewed (approved):
LGTM. Thank you, @octonawish-akcodes. Merging!
